src/caffeine/teams_s.f90

```diff
@@ -1,20 +1,70 @@
 ! Copyright (c), The Regents of the University of California
 ! Terms of use are as specified in LICENSE.txt
 submodule(prif:prif_private_s) teams_s
-  implicit none
+  use iso_c_binding, only: c_null_funptr, c_f_pointer, c_loc
 
+  implicit none
 contains
 
   module procedure prif_change_team
-    call unimplemented("prif_change_team")
+    if (caf_this_image(team%info%gex_team) == 1) then ! need to setup the heap for the team
+      team%info%heap_start = current_team%info%child_heap_info%offset + current_team%info%heap_start
+      team%info%heap_size = current_team%info%child_heap_info%size
+      call caf_establish_mspace( &
+          team%info%heap_mspace, &
+          as_c_ptr(team%info%heap_start), &
+          current_team%info%child_heap_info%size)
+    end if
+    current_team = team
+    if (caf_have_child_teams()) then ! need to establish heap for child teams
+      call caf_establish_child_heap
+    end if
   end procedure
 
   module procedure prif_end_team
-    call unimplemented("prif_end_team")
+    type(prif_coarray_handle), allocatable :: teams_coarrays(:)
+    integer :: num_coarrays_in_team, i
+    type(handle_data), pointer :: tmp_data
+
+    ! deallocate the teams coarrays
+    num_coarrays_in_team = 0
+    tmp_data => current_team%info%coarrays
+    do while (associated(tmp_data))
+      num_coarrays_in_team = num_coarrays_in_team + 1
+      call c_f_pointer(tmp_data%next_handle, tmp_data)
+    end do
+    if (num_coarrays_in_team > 0) then
+      allocate(teams_coarrays(num_coarrays_in_team))
+      tmp_data => current_team%info%coarrays
+      do i = 1, num_coarrays_in_team
+        teams_coarrays(i)%info => tmp_data
+        call c_f_pointer(tmp_data%next_handle, tmp_data)
+      end do
+      call prif_deallocate_coarray(teams_coarrays, stat, errmsg, errmsg_alloc)
```

Comment:
IIUC `current_team%info%coarrays` is a dangling pointer now and needs to be nullified to prevent an error after the next change team to this team.

```suggestion
      call prif_deallocate_coarray(teams_coarrays, stat, errmsg, errmsg_alloc)
      nullify(current_team%info%coarrays)
```

Reply:
Another good catch. I now do so.